### PR TITLE
Add extract transform mimerender

### PR DESCRIPTION
**Pull request by saulshanabrook**
This adds the ability to extract the transforms from Python vega lite specs by using `extractTransforms` from `vega-lite` JS library. This PR let's us do this from the notebook directly with specs and provides Altair renderers to have Altair call this transformation on it's results.

This sets us up to do any logic we want in Python land on the extracted spec, like updating an Ibis SQL statement with the transformations.

Check out new `./notebooks` directory for walkthroughs of the new features:

http://nbviewer.jupyter.org/github/Quansight/jupyterlab-omnisci/tree/vega-lite-transform/notebooks/

### CODE REVIEW

**Comment by ian-r-rose** [comment]
How does this interact with the existing backend mimerenderer? I see you have renamed things so they are no longer compiled. Do you view this as an alternative or a replacement?

**Comment by saulshanabrook** [comment]
This is a replacement for that workflow. You are right, since I just upgraded VL I had to comment those out since they wouldn't compile with the new version. I didn't wanna spend time fixing them unless we are moving forward on that old approach.

Old approach: use mapd's backend vega rendering to show graphs
new approach: use frontend vega lite rendering, but extract aggregates first so that we send less data to client

**Comment by saulshanabrook** [comment]
There isn't a reason besides time :) But it probably isn't much time, so I will uncomment those and try to update them to keep the code from being dead.

**Comment by ian-r-rose** [comment]
I believe the raw-vega(ish) approach still works, as I was just testing it :). I am not sure about the vega-lite stuff.

**Comment by saulshanabrook** [comment]
True, I forgot about the raw vega workflow, that is still nice.

**Comment by ian-r-rose** [comment]
I also have #31 open, which reuses the same mapd connection in several places, and speeds up some of the operations (especially tab-completion). I think it is mostly orthogonal to what you have here, but there may be some conflicts at the margins. How do you want to move forward with these?

**Comment by saulshanabrook** [comment]
Let's merge yours first, then I can rebase off that.

**Comment by saulshanabrook** [comment]
@ian-r-rose this is ready to merge from my perspective.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (saulshanabrook) [reply .vscode/settings.json]: Good catch

File: .gitignore [hunk 2]
```diff
@@ -12,3 +12,4 @@ __pycache__
 .yarnrc
 .bash_history
 .config
+.mypy_cache
```
[the remark this answers is not in the document]
Comment (saulshanabrook) [reply .gitignore]: I started adding some, but then took it out. I also find mypy sometimes catches things even without adding any annotations, just on core functions, so I find that helpful.

**Comment by ian-r-rose** [comment]
The moved notebooks make it so that they can no longer import `omnisci_renderer`. Should I think we should clean up that python file and make it an actual package (but possibly in a follow-up). In the meantime, we may want to move the notebooks back to the root.

**Comment by ian-r-rose** [comment]
Some of the demonstrations relating to vega-lite in `renderer.ipynb` have broken in this. I suppose that's to be expected. Do you have any thoughts about whether they should be updated or removed altogether?

**Comment by ian-r-rose** [comment]
When I run the notebooks, the plots seem to appear correctly, but I also get a bunch of errors on the console like
```
Exception opening new comm default.js:924
Error: "Object 'extract-vega-lite' not found in registry"
	loadObject http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:149248:28
	loadObject http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:149227:16
	_handleCommOpen http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:148693:36
	__awaiter http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:147785:71
	__awaiter http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:147781:12
	_handleCommOpen http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:148685:16
	_handleMessage http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:148792:31
	__awaiter http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:147785:71
	__awaiter http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:147781:12
	_handleMessage http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:148749:16
	_msgChain http://localhost:8888/lab/static/vendors~main.035877a516da07693300.js:147903:24
default.js:132
The compiled spec uses Vega v3, but current version is 4.3.0. embed.js:137
Error: Invalid spec
```

**Comment by saulshanabrook** [comment]
@ian-r-rose Good point about notebook locations. They should be moved back.

I will look at the things broken in `renderer.ipynb`, hopefully not a lot will break.

Did you re-install the extension?

**Comment by ian-r-rose** [comment]
In general, I am in favor of moving them to a `notebooks` directory, we should just properly packageify the python stuff.

**Comment by ian-r-rose** [comment]
Ha, you are right, looks like there was some stale JS around. After rebuilding I now only see the vega-3-vs-4 warnings.

**Comment by saulshanabrook** [comment]
OK I fixed one thing and now the two original notebooks run for me, besides one cell which might be due to change in the backend mapd server rendering.

**Comment by ian-r-rose** [comment]
Looks good to me! What do you view the next steps as being?

**Comment by ian-r-rose** [comment]
Do you know why the transformed altair output has `null` output after the chart?

**Comment by ian-r-rose** [comment]
Hmm, that `null` is actually causing notebook validation to fail on my machine. Can you reproduce?

**Comment by saulshanabrook** [comment]
@ian-r-rose Yeah I see that too on my machine. It is `null`, because we are using Atlair's hook to override the `_repr_mimebundle_` of the `Chart` object. We do that here:

https://github.com/Quansight/jupyterlab-omnisci/blob/efb450ff66cfb1289498eacb843a65fdf748d556/omnisci_renderer.py#L154-L163

I am just return some null text, but I can just return empty text instead. We are doing the actual rendering  with the display update stuff that @tonyfast showed today.

Ideally I would like to return some kind of `_repr_mimebundle` that has tells the frontend not to output it at all, not make an output cell for it. But not sure if this possible.

File: omnisci_renderer.py [hunk 3]
```diff
@@ -102,25 +101,68 @@ def omnisci_vl(line, cell):
...
+        return default_renderer(self.data)
+
+
+EMPTY_SPEC = {"data": {"values": []}, "mark": "bar"}
+
+
+def extract_vega_renderer(spec):
+    display_id = display(VegaLite(EMPTY_SPEC), display_id=True)
+    extract_spec(spec, lambda s: display_id.update(VegaLite(s)))
+    return {"text/plain": None}
```
Comment (saulshanabrook) [inline omnisci_renderer.py]:
```suggestion
    return {"text/plain": ""}
```

File: omnisci_renderer.py [hunk 4]
```diff
@@ -102,25 +101,68 @@ def omnisci_vl(line, cell):
...
+
+
+def extract_vega_renderer_json(spec):
+    display_id = display(IPython.display.JSON({}), display_id=True)
+    extract_spec(spec, lambda s: display_id.update(IPython.display.JSON(s)))
+    return {"text/plain": None}
```
Comment (saulshanabrook) [inline omnisci_renderer.py]:
```suggestion
    return {"text/plain": ""}
```

File: omnisci_renderer.py (hunk 3 above)
Comment (ian-r-rose) [reply omnisci_renderer.py]: Ooh, making use of some new GitHub features, I see :smile:

**Comment by saulshanabrook** [comment]
I used Github's fancy new suggestion feature to change those lines.

I think the next step should be to merge this in, and then I can work on integrating this pipeline with ibis.

File: omnisci_renderer.py (hunk 3 above)
Comment (saulshanabrook) [reply omnisci_renderer.py]: Yep! Seems to work pretty nicely.

**Comment by ian-r-rose** [comment]
That makes sense to me. We could think of having the transient "waiting for transform" display data to be some rich HTML that shows a loading wheel or something. But it's also possible that the transform would be quick enough that it would be more annoying than anything else.  What do you think?

Merging away, thanks @saulshanabrook!

**Comment by saulshanabrook** [comment]
@ian-r-rose yeah that makes sense conceptually. The issue is that we are actually creating two outputs, one produced by the mimerender bundle we return and one produced by `display(VegaLite(EMPTY_SPEC), display_id=True)`. We need the second one, because we pass in `display_Id` to that one so we can update it later.

Ideally, we would be able to update the first one, and not have to produce the second. But I am not sure how to do this (because we don't know what display id it has to update).

**Pull request merged**